File: Year-2025 Problem List/March 2025/2570. Merge Two 2D Arrays by Summing Values/2570. Merge Two 2D Arrays by Summing Values.py
# Problem          : 2570. Merge Two 2D Arrays by Summing Values
# Time Complexity  : O(N+M)
# Python Solution

class Solution:
    def mergeArrays(
        self, nums1: List[List[int]], nums2: List[List[int]]
    ) -> List[List[int]]:

        n, m = len(nums1), len(nums2)
        left, right = 0, 0
        result = []

        while left < n and right < m:
            if nums1[left][0] == nums2[right][0]:
                result.append([nums1[left][0], nums1[left][1] + nums2[right][1]])
                left += 1
                right += 1
            elif nums1[left][0] < nums2[right][0]:
                result.append(nums1[left])
                left += 1
            else:
                result.append(nums2[right])
                right += 1

        while left < n:
            result.append(nums1[left])
            left += 1

        while right < m:
            result.append(nums2[right])
            right += 1

        return result

File: Year-2025 Problem List/March 2025/2570. Merge Two 2D Arrays by Summing Values/2570. Merge Two 2D Arrays by Summing Values.cpp
/*
# Problem          : 2570. Merge Two 2D Arrays by Summing Values
# Time complexity: O((N1 + N2) log(N1 + N2))
# Space complexity : O(N1 + N2)
# CPP Solution
*/

class Solution {
public:
    vector<vector<int>> mergeArrays(vector<vector<int>>& nums1,
                                    vector<vector<int>>& nums2) {
        map<int, int> mp;

        for (auto nums : nums1) {
            mp[nums[0]] = nums[1];
        }

        for (auto nums : nums2) {
            mp[nums[0]] += nums[1];
        }

        vector<vector<int>> ans;
        for (auto [x, y] : mp) {
            vector<int> temp = {x, y};
            ans.push_back(temp);
        }
        return ans;
    }
};
